File: src/Demos.fs
module Demos

// ─────────────────────────────────────────────────────────────────────────────
//  Los demos vivos. Cada función devuelve el nodo ya cableado: la explicación
//  y el ejemplo son el mismo objeto, que es justo lo que hace útil este deck
//  frente a un cliente.
// ─────────────────────────────────────────────────────────────────────────────

open Browser
open Browser.Types
open Dsl

let private icono (path: string) (cls: string) =
    svg cls ("<svg viewBox='0 0 24 24' aria-hidden='true'><path d='" + path + "'/></svg>")

let private etiqueta (txt: string) = span [ Cls "demo-label"; Txt txt ] []

let private marco (titulo: string) (hijos: HTMLElement list) =
    div [ Cls "demo" ] (etiqueta titulo :: hijos)

// ── 01 · Microinteracciones ──────────────────────────────────────────────────

let private botonEstado () =
    let b = button [ Cls "btn btn-state"; At ("type", "button") ] []
    b.innerHTML <- "<span>Enviar propuesta</span>"
    let mutable ocupado = false
    b.addEventListener ("click", fun _ ->
        if not ocupado then
            ocupado <- true
            b.className <- "btn btn-state is-loading"
            b.innerHTML <- "<i class='spinner'></i><span>Enviando…</span>"
            Motion.after 1500 (fun () ->
                b.className <- "btn btn-state is-done"
                b.innerHTML <-
                    "<svg class='tick' viewBox='0 0 24 24'><path d='" + Data.Ico.check + "'/></svg><span>Propuesta enviada</span>"
                Motion.after 1900 (fun () ->
                    b.className <- "btn btn-state"
                    b.innerHTML <- "<span>Enviar propuesta</span>"
                    ocupado <- false)))
    b

let private interruptor () =
    let knob = span [ Cls "switch-knob" ] []
    let sw = button [ Cls "switch"; At ("type", "button"); At ("role", "switch"); At ("aria-checked", "true"); At ("aria-label", "Alternar tema claro y oscuro") ] [ knob ]
    let mutable oscuro = true
    sw.addEventListener ("click", fun _ ->
        oscuro <- not oscuro
        toggleCls "is-off" (not oscuro) sw
        sw.setAttribute ("aria-checked", (if oscuro then "true" else "false"))
        document.documentElement.setAttribute ("data-theme", (if oscuro then "dark" else "light")))
    sw

let microinteracciones () =
    marco "Pruébelo: pase el cursor y haga clic"
        [ div [ Cls "demo-row" ]
            [ button [ Cls "btn btn-hover-color"; At ("type", "button"); Txt "Cambia de color" ] []
              button [ Cls "btn btn-hover-grow"; At ("type", "button"); Txt "Crece 5 %" ] []
              button [ Cls "btn btn-hover-lift"; At ("type", "button"); Txt "Se eleva" ] [] ]
          div [ Cls "demo-row demo-row--split" ]
            [ botonEstado ()
              div [ Cls "switch-wrap" ]
                [ interruptor ()
                  span [ Cls "switch-txt"; Txt "Modo oscuro" ] [] ] ] ]

// ── 02 · Animaciones de entrada ──────────────────────────────────────────────

let entrada () =
    let tarjetas =
        [ "Diagnóstico", "Auditamos la interfaz actual"
          "Guion", "Definimos qué se revela y cuándo"
          "Prototipo", "Movimiento real, no maqueta"
          "Implementación", "Código listo para producción"
          "Medición", "Antes y después del comportamiento"
          "Iteración", "Se ajusta la curva, no el color" ]

    let grid =
        div [ Cls "stagger-grid" ]
            (tarjetas
             |> List.mapi (fun i (t, d) ->
                 div [ Cls "sc" ]
                     [ span [ Cls "sc-n"; Txt (sprintf "%02d" (i + 1)) ] []
                       h4 [ Txt t ] []
                       p [ Cls "sc-d"; Txt d ] [] ]))

    Motion.stagger grid 90

    let replay =
        button [ Cls "btn btn-ghost"; At ("type", "button"); Txt "Reproducir de nuevo" ] []
    replay.addEventListener ("click", fun _ -> Motion.replay grid "is-in")

    marco "Cascada de 6 tarjetas · 90 ms entre cada una"
        [ grid
          div [ Cls "demo-row" ] [ replay ] ]

// ── 03 · Scroll ──────────────────────────────────────────────────────────────

/// Bloque alto con panel fijo: la escena cambia según el % de avance.
let scrollytelling () =
    let dial = div [ Cls "scrolly-dial" ] []
    let pct = span [ Cls "scrolly-pct"; Txt "0 %" ] []
    let barra = div [ Cls "scrolly-bar" ] [ div [ Cls "scrolly-bar-fill" ] [] ]

    let escenas =
        [ "El cliente llega", "La página está quieta. Todavía no pasa nada."
          "El cliente baja", "Cada bloque aparece justo cuando toca leerlo."
          "El cliente entiende", "El movimiento ordenó la información por usted."
          "El cliente decide", "La demo terminó y el producto se explicó solo." ]

    let nodos =
        escenas
        |> List.mapi (fun i (t, d) ->
            div [ Cls (if i = 0 then "scene is-on" else "scene") ]
                [ span [ Cls "scene-n"; Txt (sprintf "0%d" (i + 1)) ] []
                  h4 [ Txt t ] []
                  p [ Cls "scene-d"; Txt d ] [] ])

    let panel =
        div [ Cls "scrolly-panel" ]
            [ div [ Cls "scrolly-visual" ] [ dial; pct ]
              div [ Cls "scrolly-text" ] nodos
              barra ]

    let sticky = div [ Cls "scrolly-sticky" ] [ panel ]
    let bloque = section [ Cls "scrolly"; Id "scrollytelling" ] [ sticky ]

    let fill = barra.querySelector ".scrolly-bar-fill" :?> HTMLElement

    Motion.scrollProgress bloque (fun p ->
        let porcentaje = int (System.Math.Round(p * 100.0))
        pct.textContent <- string porcentaje + " %"
        setTransform fill <| "scaleX(" + string (System.Math.Round(p, 4)) + ")"
        // El disco gira y cambia de tono con el avance.
        setTransform dial <| "rotate(" + string (System.Math.Round(p * 320.0, 2)) + "deg) scale(" + string (0.72 + p * 0.34) + ")"
        setVar "--p" (string (System.Math.Round(p, 4))) dial
        let activa = min (nodos.Length - 1) (int (p * float nodos.Length))
        nodos |> List.iteri (fun i n -> toggleCls "is-on" (i = activa) n))

    bloque

let scroll () =
    marco "Baje despacio: el fondo va a un tercio de su velocidad"
        [ div [ Cls "parallax-demo" ]
            [ div [ Cls "px-layer px-back" ] []
              div [ Cls "px-layer px-mid" ] []
              div [ Cls "px-copy" ]
                [ h4 [ Txt "Profundidad real" ] []
                  p [ Cls "sc-d"; Txt "Tres capas a distinta velocidad. El texto siempre se lee." ] [] ] ]
          para "demo-note" "El scrollytelling completo está justo debajo de esta sección." ]

// ── 04 · Cursor ──────────────────────────────────────────────────────────────

let private botonMagnetico () =
    let inner = span [ Cls "mag-inner"; Txt "Acérquese" ] []
    let wrap = div [ Cls "mag" ] [ inner ]
    Motion.magnetic wrap inner 130.0 0.38
    wrap

let private cara () =
    let hacerOjo () =
        let pup = span [ Cls "pupila" ] []
        let ojo = div [ Cls "ojo" ] [ pup ]
        ojo, pup

    let o1, p1 = hacerOjo ()
    let o2, p2 = hacerOjo ()
    let c = div [ Cls "cara" ] [ o1; o2 ]
    Motion.pupila o1 p1 13.0
    Motion.pupila o2 p2 13.0
    Motion.follow c 10.0
    c

let cursor () =
    marco "Mueva el mouse por esta tarjeta"
        [ div [ Cls "demo-row demo-row--split cursor-row" ]
            [ botonMagnetico ()
              cara () ]
          para "demo-note" "En pantalla táctil el efecto no se monta: entra la microinteracción de toque." ]

// ── 05 · Media enriquecida ───────────────────────────────────────────────────

let media () =
    let host = div [ Cls "lottie-host"; At ("aria-live", "polite") ] []
    let estado = span [ Cls "chip chip-live"; Txt "en espera" ] []
    let peso = span [ Cls "chip"; Txt "midiendo…" ] []

    Lottie.mount host "./assets/lottie/pulse.json" (fun s -> estado.textContent <- s)
    Lottie.pesaJson "./assets/lottie/pulse.json" (fun bytes ->
        if bytes = 0 then peso.textContent <- "peso no disponible"
        else
            let kb = float bytes / 1024.0
            peso.textContent <- "JSON real: " + string (System.Math.Round(kb, 1)) + " KB")

    let comparativa =
        div [ Cls "vs" ]
            [ div [ Cls "vs-row" ]
                [ span [ Cls "vs-k"; Txt "Lottie (JSON vectorial)" ] []
                  div [ Cls "vs-bar" ] [ div [ Cls "vs-fill vs-a"; Sty "width:4%" ] [] ]
                  span [ Cls "vs-v"; Txt "≈ 4 KB" ] [] ]
              div [ Cls "vs-row" ]
                [ span [ Cls "vs-k"; Txt "GIF equivalente 600×600" ] []
                  div [ Cls "vs-bar" ] [ div [ Cls "vs-fill vs-b"; Sty "width:100%" ] [] ]
                  span [ Cls "vs-v"; Txt "≈ 450 KB" ] [] ]
              para "demo-note"
                  "Además el GIF se pixela al escalar y no se puede recolorear. El Lottie es vectorial, \
                   se reproduce a cualquier tamaño y su color se cambia desde código." ]

    let lienzo = canvas [ Cls "particles"; At ("aria-hidden", "true") ] []
    Particles.mount lienzo 90

    marco "Dos técnicas de alto nivel, funcionando"
        [ div [ Cls "media-grid" ]
            [ div [ Cls "media-card" ]
                [ host
                  div [ Cls "chips" ] [ estado; peso ]
                  comparativa ]
              div [ Cls "media-card media-card--canvas" ]
                [ lienzo
                  div [ Cls "canvas-cap" ]
                    [ h4 [ Txt "Campo de partículas 3D" ] []
                      para "sc-d"
                          "Proyección en perspectiva sobre Canvas. El cursor inclina la cámara y \
                           empuja las partículas cercanas." ] ] ] ]

// ── Selector ─────────────────────────────────────────────────────────────────

let render (d: Data.Demo) =
    match d with
    | Data.DemoMicro -> microinteracciones ()
    | Data.DemoEntrada -> entrada ()
    | Data.DemoScroll -> scroll ()
    | Data.DemoCursor -> cursor ()
    | Data.DemoMedia -> media ()

File: paginas/motion-system/scripts/Demos.js
import { canvas, para, setVar, section, p as p_1, h4, toggleCls, button, div, Attr$, span, svg } from "./Dsl.js";
import { iterateIndexed, length, mapIndexed, cons, singleton, empty, ofArray } from "./fable_modules/fable-library-js.4.24.0/List.js";
import { follow, pupila, magnetic, scrollProgress, replay as replay_1, stagger, after } from "./Motion.js";
import { Ico_check } from "./Data.js";
import { printf, toText } from "./fable_modules/fable-library-js.4.24.0/String.js";
import { int32ToString, round } from "./fable_modules/fable-library-js.4.24.0/Util.js";
import { min } from "./fable_modules/fable-library-js.4.24.0/Double.js";
import { mount } from "./Lottie.js";
import { mount as mount_1 } from "./Particles.js";

function icono(path, cls) {
    return svg(cls, ("<svg viewBox=\'0 0 24 24\' aria-hidden=\'true\'><path d=\'" + path) + "\'/></svg>");
}

function etiqueta(txt) {
    return span(ofArray([new Attr$(0, ["demo-label"]), new Attr$(4, [txt])]), empty());
}

function marco(titulo, hijos) {
    return div(singleton(new Attr$(0, ["demo"])), cons(etiqueta(titulo), hijos));
}

function botonEstado() {
    const b = button(ofArray([new Attr$(0, ["btn btn-state"]), new Attr$(2, ["type", "button"])]), empty());
    b.innerHTML = "<span>Enviar propuesta</span>";
    let ocupado = false;
    b.addEventListener("click", (_arg) => {
        if (!ocupado) {
            ocupado = true;
            b.className = "btn btn-state is-loading";
            b.innerHTML = "<i class=\'spinner\'></i><span>Enviando…</span>";
            after(1500, () => {
                b.className = "btn btn-state is-done";
                b.innerHTML = (("<svg class=\'tick\' viewBox=\'0 0 24 24\'><path d=\'" + Ico_check) + "\'/></svg><span>Propuesta enviada</span>");
                after(1900, () => {
                    b.className = "btn btn-state";
                    b.innerHTML = "<span>Enviar propuesta</span>";
                    ocupado = false;
                });
            });
        }
    });
    return b;
}

function interruptor() {
    const sw = button(ofArray([new Attr$(0, ["switch"]), new Attr$(2, ["type", "button"]), new Attr$(2, ["role", "switch"]), new Attr$(2, ["aria-checked", "true"]), new Attr$(2, ["aria-label", "Alternar tema claro y oscuro"])]), singleton(span(singleton(new Attr$(0, ["switch-knob"])), empty())));
    let oscuro = true;
    sw.addEventListener("click", (_arg) => {
        oscuro = !oscuro;
        toggleCls("is-off", !oscuro, sw);
        sw.setAttribute("aria-checked", oscuro ? "true" : "false");
        document.documentElement.setAttribute("data-theme", oscuro ? "dark" : "light");
    });
    return sw;
}

export function microinteracciones() {
    return marco("Pruébelo: pase el cursor y haga clic", ofArray([div(singleton(new Attr$(0, ["demo-row"])), ofArray([button(ofArray([new Attr$(0, ["btn btn-hover-color"]), new Attr$(2, ["type", "button"]), new Attr$(4, ["Cambia de color"])]), empty()), button(ofArray([new Attr$(0, ["btn btn-hover-grow"]), new Attr$(2, ["type", "button"]), new Attr$(4, ["Crece 5 %"])]), empty()), button(ofArray([new Attr$(0, ["btn btn-hover-lift"]), new Attr$(2, ["type", "button"]), new Attr$(4, ["Se eleva"])]), empty())])), div(singleton(new Attr$(0, ["demo-row demo-row--split"])), ofArray([botonEstado(), div(singleton(new Attr$(0, ["switch-wrap"])), ofArray([interruptor(), span(ofArray([new Attr$(0, ["switch-txt"]), new Attr$(4, ["Modo oscuro"])]), empty())]))]))]));
}

export function entrada() {
    const grid = div(singleton(new Attr$(0, ["stagger-grid"])), mapIndexed((i, tupledArg) => {
        let arg;
        return div(singleton(new Attr$(0, ["sc"])), ofArray([span(ofArray([new Attr$(0, ["sc-n"]), new Attr$(4, [(arg = ((i + 1) | 0), toText(printf("%02d"))(arg))])]), empty()), h4(singleton(new Attr$(4, [tupledArg[0]])), empty()), p_1(ofArray([new Attr$(0, ["sc-d"]), new Attr$(4, [tupledArg[1]])]), empty())]));
    }, ofArray([["Diagnóstico", "Auditamos la interfaz actual"], ["Guion", "Definimos qué se revela y cuándo"], ["Prototipo", "Movimiento real, no maqueta"], ["Implementación", "Código listo para producción"], ["Medición", "Antes y después del comportamiento"], ["Iteración", "Se ajusta la curva, no el color"]])));
    stagger(grid, 90);
    const replay = button(ofArray([new Attr$(0, ["btn btn-ghost"]), new Attr$(2, ["type", "button"]), new Attr$(4, ["Reproducir de nuevo"])]), empty());
    replay.addEventListener("click", (_arg) => {
        replay_1(grid, "is-in");
    });
    return marco("Cascada de 6 tarjetas · 90 ms entre cada una", ofArray([grid, div(singleton(new Attr$(0, ["demo-row"])), singleton(replay))]));
}

/**
 * Bloque alto con panel fijo: la escena cambia según el % de avance.
 */
export function scrollytelling() {
    const dial = div(singleton(new Attr$(0, ["scrolly-dial"])), empty());
    const pct = span(ofArray([new Attr$(0, ["scrolly-pct"]), new Attr$(4, ["0 %"])]), empty());
    const barra = div(singleton(new Attr$(0, ["scrolly-bar"])), singleton(div(singleton(new Attr$(0, ["scrolly-bar-fill"])), empty())));
    const nodos = mapIndexed((i, tupledArg) => {
        let arg;
        return div(singleton(new Attr$(0, [(i === 0) ? "scene is-on" : "scene"])), ofArray([span(ofArray([new Attr$(0, ["scene-n"]), new Attr$(4, [(arg = ((i + 1) | 0), toText(printf("0%d"))(arg))])]), empty()), h4(singleton(new Attr$(4, [tupledArg[0]])), empty()), p_1(ofArray([new Attr$(0, ["scene-d"]), new Attr$(4, [tupledArg[1]])]), empty())]));
    }, ofArray([["El cliente llega", "La página está quieta. Todavía no pasa nada."], ["El cliente baja", "Cada bloque aparece justo cuando toca leerlo."], ["El cliente entiende", "El movimiento ordenó la información por usted."], ["El cliente decide", "La demo terminó y el producto se explicó solo."]]));
    const bloque = section(ofArray([new Attr$(0, ["scrolly"]), new Attr$(1, ["scrollytelling"])]), singleton(div(singleton(new Attr$(0, ["scrolly-sticky"])), singleton(div(singleton(new Attr$(0, ["scrolly-panel"])), ofArray([div(singleton(new Attr$(0, ["scrolly-visual"])), ofArray([dial, pct])), div(singleton(new Attr$(0, ["scrolly-text"])), nodos), barra]))))));
    const fill = barra.querySelector(".scrolly-bar-fill");
    scrollProgress(bloque, (p) => {
        const porcentaje = ~~round(p * 100) | 0;
        pct.textContent = (int32ToString(porcentaje) + " %");
        const v = ("scaleX(" + round(p, 4).toString()) + ")";
        fill.style.transform = v;
        const v_1 = ((("rotate(" + round(p * 320, 2).toString()) + "deg) scale(") + (0.72 + (p * 0.34)).toString()) + ")";
        dial.style.transform = v_1;
        setVar("--p", round(p, 4).toString(), dial);
        const activa = min(length(nodos) - 1, ~~(p * length(nodos))) | 0;
        iterateIndexed((i_1, n) => {
            toggleCls("is-on", i_1 === activa, n);
        }, nodos);
    });
    return bloque;
}

export function scroll() {
    return marco("Baje despacio: el fondo va a un tercio de su velocidad", ofArray([div(singleton(new Attr$(0, ["parallax-demo"])), ofArray([div(singleton(new Attr$(0, ["px-layer px-back"])), empty()), div(singleton(new Attr$(0, ["px-layer px-mid"])), empty()), div(singleton(new Attr$(0, ["px-copy"])), ofArray([h4(singleton(new Attr$(4, ["Profundidad real"])), empty()), p_1(ofArray([new Attr$(0, ["sc-d"]), new Attr$(4, ["Tres capas a distinta velocidad. El texto siempre se lee."])]), empty())]))])), para("demo-note", "El scrollytelling completo está justo debajo de esta sección.")]));
}

function botonMagnetico() {
    const inner = span(ofArray([new Attr$(0, ["mag-inner"]), new Attr$(4, ["Acérquese"])]), empty());
    const wrap = div(singleton(new Attr$(0, ["mag"])), singleton(inner));
    magnetic(wrap, inner, 130, 0.38);
    return wrap;
}

function cara() {
    const hacerOjo = () => {
        const pup = span(singleton(new Attr$(0, ["pupila"])), empty());
        return [div(singleton(new Attr$(0, ["ojo"])), singleton(pup)), pup];
    };
    const patternInput = hacerOjo();
    const o1 = patternInput[0];
    const patternInput_1 = hacerOjo();
    const o2 = patternInput_1[0];
    const c = div(singleton(new Attr$(0, ["cara"])), ofArray([o1, o2]));
    pupila(o1, patternInput[1], 13);
    pupila(o2, patternInput_1[1], 13);
    follow(c, 10);
    return c;
}

export function cursor() {
    return marco("Mueva el mouse por esta tarjeta", ofArray([div(singleton(new Attr$(0, ["demo-row demo-row--split cursor-row"])), ofArray([botonMagnetico(), cara()])), para("demo-note", "En pantalla táctil el efecto no se monta: entra la microinteracción de toque.")]));
}

export function media() {
    const host = div(ofArray([new Attr$(0, ["lottie-host"]), new Attr$(2, ["aria-live", "polite"])]), empty());
    const estado = span(ofArray([new Attr$(0, ["chip chip-live"]), new Attr$(4, ["en espera"])]), empty());
    const peso = span(ofArray([new Attr$(0, ["chip"]), new Attr$(4, ["midiendo…"])]), empty());
    mount(host, "../../compartidos/lottie/pulse.json", (s) => {
        estado.textContent = s;
    });
    fetch("../../compartidos/lottie/pulse.json").then(function(r){return r.text();}).then(function(t){((bytes) => {
        if (bytes === 0) {
            peso.textContent = "peso no disponible";
        }
        else {
            const kb = bytes / 1024;
            peso.textContent = (("JSON real: " + round(kb, 1).toString()) + " KB");
        }
    })(new Blob([t]).size);}).catch(function(){((bytes) => {
        if (bytes === 0) {
            peso.textContent = "peso no disponible";
        }
        else {
            const kb = bytes / 1024;
            peso.textContent = (("JSON real: " + round(kb, 1).toString()) + " KB");
        }
    })(0);});
    const comparativa = div(singleton(new Attr$(0, ["vs"])), ofArray([div(singleton(new Attr$(0, ["vs-row"])), ofArray([span(ofArray([new Attr$(0, ["vs-k"]), new Attr$(4, ["Lottie (JSON vectorial)"])]), empty()), div(singleton(new Attr$(0, ["vs-bar"])), singleton(div(ofArray([new Attr$(0, ["vs-fill vs-a"]), new Attr$(3, ["width:4%"])]), empty()))), span(ofArray([new Attr$(0, ["vs-v"]), new Attr$(4, ["≈ 4 KB"])]), empty())])), div(singleton(new Attr$(0, ["vs-row"])), ofArray([span(ofArray([new Attr$(0, ["vs-k"]), new Attr$(4, ["GIF equivalente 600×600"])]), empty()), div(singleton(new Attr$(0, ["vs-bar"])), singleton(div(ofArray([new Attr$(0, ["vs-fill vs-b"]), new Attr$(3, ["width:100%"])]), empty()))), span(ofArray([new Attr$(0, ["vs-v"]), new Attr$(4, ["≈ 450 KB"])]), empty())])), para("demo-note", "Además el GIF se pixela al escalar y no se puede recolorear. El Lottie es vectorial, se reproduce a cualquier tamaño y su color se cambia desde código.")]));
    const lienzo = canvas(ofArray([new Attr$(0, ["particles"]), new Attr$(2, ["aria-hidden", "true"])]), empty());
    mount_1(lienzo, 90);
    return marco("Dos técnicas de alto nivel, funcionando", singleton(div(singleton(new Attr$(0, ["media-grid"])), ofArray([div(singleton(new Attr$(0, ["media-card"])), ofArray([host, div(singleton(new Attr$(0, ["chips"])), ofArray([estado, peso])), comparativa])), div(singleton(new Attr$(0, ["media-card media-card--canvas"])), ofArray([lienzo, div(singleton(new Attr$(0, ["canvas-cap"])), ofArray([h4(singleton(new Attr$(4, ["Campo de partículas 3D"])), empty()), para("sc-d", "Proyección en perspectiva sobre Canvas. El cursor inclina la cámara y empuja las partículas cercanas.")]))]))]))));
}

export function render(d) {
    switch (d.tag) {
        case 1:
            return entrada();
        case 2:
            return scroll();
        case 3:
            return cursor();
        case 4:
            return media();
        default:
            return microinteracciones();
    }
}

